### PR TITLE
Feature/eeprom

### PR DESCRIPTION
**Pull request by wvarty**

### CODE REVIEW

**Comment by wvarty** [comment]
This will need a thorough test before merging.

**Comment by wvarty** [comment]
One thing I forgot to add to the eeprom struct was a CRC... I havent used one in past projects, and it's been fine, but I noticed that the chorus32 eeprom struct uses one, and I think when I last spoke to @AlessandroAU and @JyeSmith about this, they were pretty firm on needing it?

**Comment by JyeSmith** [comment]
@wvarty previously I have only used the version number as a check.  Worse case (or maybe best) is that it is corrupt and the UIDs are different so the quad wont bind.

image

**Comment by wvarty** [comment]
@JyeSmith I guess worst case is prob more like we read a corrupt value for power output, and someone goes to fly freestyle thinking they are on 200mw, but are actually on 25mw, and totals a quad...

**Comment by JBKingdon** [comment]
How often do we expect the values to be written, do we need to worry about wearing out the flash?

**Comment by wvarty** [comment]
Everytime we move the scroll wheel in LUA, it will do a write... I dislike how we send the value change to the TX from the handset on scroll... IMO it should only get sent when you press the return button

**Comment by wvarty** [comment]
The only way I can think of to reduce write cycles is to put some hysteresis around detecting changes... I.e in the main loop you have a func that runs every 10s to see if the current vals != saved vals, and do the write. Prob with that is, you might have already armed by that time, and we need to stop the timer to do a write.

**Comment by maybenikhil** [comment]
@wvarty what if there was a button that saved (could be put in place of the bind button for now)

**Comment by JBKingdon** [comment]
Looks awesome. I like Nikhil's idea for a manual save UI element ("save as defaults"), but we also talked about auto save on exit from the lua script which would be more convenient in a lot of cases.

**Comment by JyeSmith** [comment]
> 
> 
> Everytime we move the scroll wheel in LUA, it will do a write... I dislike how we send the value change to the TX from the handset on scroll... IMO it should only get sent when you press the return button

This gets my vote.

File: src/src/config.h
Comment (JyeSmith) [inline src/src/config.h:8]: Already defined in elrs_eeprom.h.  Should it still be here?

**Comment by wvarty** [comment]
Horribly broken on STM :( dont merge

File: src/lib/EEPROM/elrs_eeprom.cpp
Comment (JBKingdon) [inline src/lib/EEPROM/elrs_eeprom.cpp:45]: We probably don't want per-byte commit. Can we move the commits up to a transactional level so that there's just one per save?

**Comment by wvarty** [comment]
Ready for another round of review. Changes were:
- Changed templates to void, and removed the ref symbol
- Added eeprom address to STM begin()
- Split out the Commit() call as per James's suggestion, and only call after all bytes have been written.
- Increased the reserved eeprom size from 512 to 2048 **NOTE: I HAVE NO IDEA WHAT THIS SHOULD ACTUALLY BE??**

File: src/src/config.cpp
Comment (AlessandroAU) [inline src/src/config.cpp:64]: we don't actually need ICACHE_RAM_ATTR unless we are calling from ISR

**Comment by JBKingdon** [comment]
As discussed, just a reminder that it would be great if we could use the put & get templates from the esp32 impl as then we would have a consistent API on the two platforms plus benefit from the 'don't write if the data hasn't changed' behaviour.

File: src/lib/EEPROM/elrs_eeprom.cpp
Comment (JyeSmith) [inline src/lib/EEPROM/elrs_eeprom.cpp:5]: IIRC the r9mm uses the default address 0x50.  Something to keep in mind when reusing this lib for the receivers.

File: src/lib/EEPROM/elrs_eeprom.cpp
Comment (JyeSmith) [inline src/lib/EEPROM/elrs_eeprom.cpp:13]: We may need to keep set and begin.  The r9 receivers do not use the DAC lib where Wire is initialised.  I dont think there is any problem with calling begin twice, but needs to be checked.

**Comment by JyeSmith** [comment]
Bench tested with r9 and esp 2.4 gear.  No probs.

File: src/lib/EEPROM/elrs_eeprom.cpp
Comment (JBKingdon) [reply src/lib/EEPROM/elrs_eeprom.cpp:53]: sweet

File: src/src/config.h
Comment (JBKingdon) [inline src/src/config.h:15]: uh, do we want the modified flag to be written to the flash? This feels like it should be just a runtime thing that would be a seperate bool in the class

Comment (wvarty) [reply src/src/config.h:15]: Good point. This should be split out.

**Comment by JBKingdon** [comment]
Looks good. Maybe a bit short of doc/comments but the code is pretty readable.

File: src/src/config.h
Comment (wvarty) [reply src/src/config.h:8]: No, it should not.

**Comment by wvarty** [comment]
> Looks good. Maybe a bit short of doc/comments but the code is pretty readable.

Hmm, I thought it was fairly comment rich... I have no problems adding more if you think it will be useful to others reading this code later?

**Comment by JBKingdon** [comment]
> Hmm, I thought it was fairly comment rich... I have no problems adding more if you think it will be useful to others reading this code later?

So, not sure what I was looking at now that made me say that. A quick look again shows it's fine. Sorry.

File: src/src/config.h
Comment (JyeSmith) [inline src/src/config.h:9]: Because a struct is being written and we don't need to track addresses, the variable types can be changed easily in the struct.  Is it worth while dropping them down to uint8_t and reduce the commit time?

**Review by JyeSmith — APPROVED** [review]

**Comment by JyeSmith** [comment]
Tested r9m to r9mm and esp 2.4 gear.  No probs with this PR.

Nice work :fireworks: :fireworks: :fireworks:

**Comment by raffxj** [comment]
guys

not working on mine. 'FW update error.no sync'

thanks 👍

File: src/src/config.h
Comment (wvarty) [reply src/src/config.h:9]: Replying here for completeness: I don't think its worth a millisecond or two in reducing these types, and it will likely cause pain later when we wanna bump up the size again, and would have to wipe people's settings as part of doing so.

**Pull request merged**
